src/mito2/src/sst/index/creator.rs

```diff
@@ -0,0 +1,264 @@
+
+            value_buf: vec![],
+
+            stats: Statistics::default(),
+            aborted: false,
+        }
+    }
+
+    /// Updates index with a batch of rows.
+    /// Garbage will be cleaned up if failed to update.
+    pub async fn update(&mut self, batch: &Batch) -> Result<()> {
+        ensure!(!self.aborted, OperateAbortedIndexSnafu);
+
+        if batch.is_empty() {
+            return Ok(());
+        }
+
+        if let Err(update_err) = self.do_update(batch).await {
+            // clean up garbage if failed to update
+            if let Err(err) = self.do_cleanup().await {
+                warn!(
+                    err; "Failed to clean up index creator, region_dir: {}, sst_file_id: {}",
+                    self.region_dir, self.sst_file_id,
+                );
+            }
+            return Err(update_err);
+        }
+
+        Ok(())
+    }
+
+    /// Finishes index creation and cleans up garbage.
+    /// Returns the number of rows and bytes written.
+    pub async fn finish(&mut self) -> Result<(RowCount, ByteCount)> {
+        ensure!(!self.aborted, OperateAbortedIndexSnafu);
+
+        if self.stats.row_count() == 0 {
+            // no IO is performed, no garbage to clean up, just return
+            return Ok((0, 0));
+        }
+
+        let finish_res = self.do_finish().await;
+        // clean up garbage no matter finish successfully or not
+        if let Err(err) = self.do_cleanup().await {
+            warn!(
+                err; "Failed to clean up index creator, region_dir: {}, sst_file_id: {}",
+                self.region_dir, self.sst_file_id,
+            );
+        }
+
+        finish_res.map(|_| (self.stats.row_count(), self.stats.byte_count()))
+    }
+
+    /// Aborts index creation and clean up garbage.
+    pub async fn abort(&mut self) -> Result<()> {
+        if self.aborted {
+            return Ok(());
+        }
+        self.aborted = true;
+
+        self.do_cleanup().await
+    }
+
+    async fn do_update(&mut self, batch: &Batch) -> Result<()> {
+        let mut guard = self.stats.record_update();
+
+        let n = batch.num_rows();
+        guard.inc_row_count(n);
+
+        for (column_name, field, value) in self.codec.decode(batch.primary_key())? {
+            if let Some(value) = value.as_ref() {
+                self.value_buf.clear();
+                IndexValueCodec::encode_value(value.as_value_ref(), field, &mut self.value_buf)?;
+            }
+
+            // non-null value -> Some(encoded_bytes), null value -> None
+            let value = value.is_some().then_some(self.value_buf.as_slice());
+            self.index_creator
+                .push_with_name_n(column_name, value, n)
+                .await
+                .context(PushIndexValueSnafu)?;
+        }
+
+        Ok(())
+    }
+
+    async fn do_finish(&mut self) -> Result<()> {
+        let mut guard = self.stats.record_finish();
+
+        // Data flow of finishing index:
```

Comment:
Is it better to put this as a document comment of `do_finish` which can be rendered as markdown?